keras_cv/layers/object_detection3d/voxelization.py

```diff
@@ -0,0 +1,267 @@
+
+
+class Voxelization(tf.keras.layers.Layer):
+    """Dynamic voxelization and pool layer."""
+
+    def __init__(
+        self,
+        point_net: tf.keras.layers.Layer,
+        voxel_size: Sequence[float],
+        spatial_size: Sequence[float],
+        name: str = "voxelization",
+        **kwargs,
+    ):
+        super().__init__(name=name, **kwargs)
+        self._point_net = point_net
+        self._voxelization_layer = PointToVoxel(
+            voxel_size=voxel_size, spatial_size=spatial_size
+        )
+        self._voxel_size = voxel_size
+        self._spatial_size = spatial_size
+        self._voxel_spatial_size = compute_voxel_spatial_size(
+            spatial_size, self._voxel_size
+        )
+        self._voxel_spatial_size_volume = np.prod(self._voxel_spatial_size).item()
+
+    def call(
+        self,
+        point_xyz: tf.Tensor,
+        point_feature: tf.Tensor,
+        point_mask: tf.Tensor,
+        training: bool,
+    ) -> tf.Tensor:
+        """Voxelizes and learns voxel features with a point net.
+
+        B: batch_size.
+        N: number of points.
+        dim: the input dimension.
+
+        Args:
+          point_xyz: [B, N, dim] point xyz in global coordinate.
+          point_feature: [B, N, ?] point feature inputs transformed after MLP (PointNet).
+          point_mask: [B, N] valid point mask.
+          training: whether it is in training mode.
+
+        Returns:
+          voxel_feature: [B, x_max, y_max, {z_max,}, mlp_dimension] voxel
+            features. If z_max is 1, z-dim is squeezed.
+        """
+        (
+            point_voxel_feature,
+            point_voxel_id,
+            point_voxel_mask,
+        ) = self._voxelization_layer(point_xyz=point_xyz, point_mask=point_mask)
+        # TODO(tanzhenyu): move compute_point_voxel_id to here, so PointToVoxel layer is more generic.
+        point_feature = tf.concat([point_feature, point_voxel_feature], axis=-1)
+        batch_size = point_feature.shape.as_list()[0] or tf.shape(point_feature)[0]
+        # [B, N, 1]
+        point_mask_float = tf.cast(point_voxel_mask, point_feature.dtype)[
+            ..., tf.newaxis
+        ]
+        # [B, N, dim]
+        point_feature = point_feature * point_mask_float
+        point_feature = self._point_net(
+            point_feature, mask=point_mask, training=training
+        )
+        # [B, N, new_dim]
+        point_feature = point_feature * point_mask_float
+        new_dim = point_feature.shape.as_list()[-1]
+        point_feature = tf.reshape(point_feature, [-1, new_dim])
+        point_voxel_id = tf.reshape(point_voxel_id, [-1])
+        # [B * num_voxels, new_dim]
+        voxel_feature = tf.math.unsorted_segment_max(
+            point_feature, point_voxel_id, batch_size * self._voxel_spatial_size_volume
+        )
+        # unsorted_segment_max sets empty values to -inf(float).
+        voxel_feature_valid_mask = voxel_feature > -1000
```

Comment:
Maybe -1000 should be a static constant like epsilon?

Reply:
Good point. Done